pre startup checks: log if sysctl variables values need to be changed

sriduth opened the pull request:
Implement checking for some os configuration
  variables on startup.

  Check:
  - sysctl variable: /proc/sys/fs/inotify/max_user_watches
  - ulimit

  Fixes: #7130

Signed-off-by: Sriduth Jayhari <sriduth.jayhari@gmail.com>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Description: Print a log to stderr if sysctl variables (ulimit, number of watchers) needs to be changed. 
Risk Level: Low
Testing: - 
Docs Changes: -
Release Notes: -
[Optional Fixes #Issue] #7130
[Optional Deprecated:]

jmarantz commented:
please re-assign me when this is ready for review. thanks!

sriduth commented:
@jmarantz sure!

sriduth commented:
Hello @jmarantz ! I've just fixed some code format issues.

I've been wondering if it makes more sense to make the documentation of the values for various sysctl variables more prominent and implementing the checks elsewhere (maybe the python starter script?)

jmarantz commented on source/common/filesystem/inotify/watcher_impl.cc:
```diff
@@ -24,7 +24,7 @@ WatcherImpl::WatcherImpl(Event::Dispatcher& dispatcher)
           },
           Event::FileTriggerType::Edge, Event::FileReadyType::Read)) {
-  RELEASE_ASSERT(inotify_fd_ >= 0, "");
+  RELEASE_ASSERT(inotify_fd_ >= 0, "Consider increasing value of user.max_inotify_watches");
```
maybe add "via sysctl" into the help text?

jmarantz commented on source/exe/main_common.cc:
```diff
@@ -86,6 +89,8 @@ MainCommonBase::MainCommonBase(const OptionsImpl& options, Event::TimeSystem& ti
     break;
   }
+
+  check_platform_settings(file_system);
```
checkPlatformSettings(file_system)

Please read over https://github.com/envoyproxy/envoy/blob/master/STYLE.md for conventions.

jmarantz commented on source/exe/posix/platform_checks.h:
```diff
@@ -0,0 +1,43 @@
+namespace Envoy {
+
+static void check_sysctl_long(std::string path, long expected, Filesystem::Instance& file_system) {
+  if (file_system.fileExists(path)) {
```
can we put these impls into a .cc file, rather than inlining them always?

Also it looks like check_sysctl_long is only used by check_platform_settings, so it should be hidden in a .cc file.

jmarantz commented on source/exe/posix/platform_checks.h:
```diff
@@ -0,0 +1,43 @@
+
+namespace Envoy {
+
+static void check_sysctl_long(std::string path, long expected, Filesystem::Instance& file_system) {
```
Pass `path` by `absl::string_view`. If there was a good reason to pass by `std::string` then you'd use a const ref.

Pass 'expected' as an int64_t. Use absl::SimpleAtoI<int64_t> for the conversion.

jmarantz commented on source/exe/posix/platform_checks.h:
```diff
@@ -0,0 +1,43 @@
+
+static void check_sysctl_long(std::string path, long expected, Filesystem::Instance& file_system) {
+  if (file_system.fileExists(path)) {
+    long current = std::stol(file_system.fileReadToEnd(path));
```
you need to handle failures in two ways:

1. What happens if fileReadToEnd fails?
2. What happens if the string returned from that is not convertible to an int64_t.

jmarantz commented on source/exe/posix/platform_checks.h:
```diff
@@ -0,0 +1,43 @@
+    long current = std::stol(file_system.fileReadToEnd(path));
+
+    if (current < expected) {
+      std::cerr << fmt::format("{} is set to {} please consider increasing it.", current, path)
```
indicate out how to increase it, to make the message actionable by an Envoy operator.

Also you should use `ENVOY_LOG_MISC(error, ...` here.

jmarantz commented on source/exe/posix/platform_checks.h:
```diff
@@ -0,0 +1,43 @@
+
+  if (getrlimit(RLIMIT_NOFILE, &current_limits) == 0) {
+    if (current_limits.rlim_max < 102400) {
+      std::cerr << fmt::format("It looks like envoy can open only upto {} files, which is a low "
```
s/upto/up to/

Use `ENVOY_LOG_MISC`.

jmarantz commented on source/exe/win32/platform_checks.h:
```diff
@@ -0,0 +1,15 @@
+// Check system wide configurations such as user level
+// file watches, file limit permissions etc...
+static void check_platform_settings(Filesystem::Instance&) {
+  std::cerr << "platform specific checks are not implemented for win32 - please ensure that "
```
I'm not a fan of having Envoy spew this out unconditionally...maybe leave a TODO for the windows porting folks here though?

sriduth commented:
Hey @jmarantz I've implemented the fixes

- [x] filesystem error handling
- [x] string to int conversion error handling
- [x] style fixes and coding convention 

I've also changed the log lines to also mention the suggested (?) value of the parameter.
The current threshold values are taken from the envoy docs for file hard limit (https://www.envoyproxy.io/docs/envoy/latest/operations/hot_restarter)

Limit for file watches is from the issue #7130

sriduth commented:
Few more tasks:

* [ ] Resolve merge conflicts
* [ ] Make changes in BUILD to be more consistent, using a single header and implementations in .cc
      for win32 and posix (like how platform_impl_lib does it)

If OK - 
* [ ] Release note

jmarantz commented on source/exe/posix/platform_checks.cc:
```diff
@@ -0,0 +1,65 @@
+
+        return;
+      }
+    } catch (const EnvoyException& e) {
```
let's capture the exception what() in a string and log it.

jmarantz commented on source/exe/posix/platform_checks.cc:
```diff
@@ -0,0 +1,65 @@
+                         current, // 1
+                         expected // 2
+          );
+        }
```
can we log an error if we can't parse the string?

jmarantz commented on source/exe/posix/platform_checks.cc:
```diff
@@ -0,0 +1,65 @@
+void checkPlatformSettings(Filesystem::Instance& file_system) {
+  struct rlimit current_limits;
+
+  int64_t min_open_files = 102400;
```
const x2

jmarantz commented on source/exe/posix/platform_checks.cc:
```diff
@@ -0,0 +1,65 @@
+                     current_limits.rlim_max, // 0
+                     min_open_files);         // 1
+    }
+  }
```
print an error if getrlimit fails. In particular, including strerror(errno).

jmarantz commented on source/exe/win32/platform_checks.cc:
```diff
@@ -0,0 +1,17 @@
+// TODO(windows maintainers)
+void checkPlatformSettings(Filesystem::Instance&) {
+  ENVOY_LOG_MISC(
+      warn,
```
I think I mentioned this before, but i think this will be too noisy for Windows users. A TODO here should be fine (maybe an open bug too).

jmarantz commented:
/wait

jmarantz commented on source/exe/platform_checks.h:
```diff
@@ -0,0 +1,9 @@
+#include "envoy/filesystem/filesystem.h"
+
+namespace Envoy {
+
```
nit: I'd probably add a namespace 'Platform' level here.

jmarantz commented on source/exe/posix/platform_checks.cc:
```diff
@@ -0,0 +1,76 @@
+      if (absl::SimpleAtoi(file_system.fileReadToEnd(path), &current)) {
+        if (current < expected) {
+          ENVOY_LOG_MISC(error,
+                         "{0} is set to {1} please consider increasing it using `sysctl`. Envoy "
```
nit: you can just use {} rather than numeric ones here; positional is OK.

jmarantz commented on source/exe/posix/platform_checks.cc:
```diff
@@ -0,0 +1,76 @@
+      } else {
+        ENVOY_LOG_MISC(error, "Could not read contents of {0} as a number",
+                       "Envoy will log this message if the value could not be read as an int or is "
+                       "less than {1}",
```
positional is OK.

jmarantz commented on source/exe/posix/platform_checks.cc:
```diff
@@ -0,0 +1,76 @@
+        );
+      }
+    } catch (const EnvoyException& e) {
+      ENVOY_LOG_MISC(error, "Exception when trying to check value of {} - {}", path, e.what());
```
s/value/contents/ ?

jmarantz commented on source/exe/win32/platform_checks.cc:
```diff
@@ -0,0 +1,12 @@
+
+// Check system wide configurations such as user level
+// file watches, file limit permissions etc...
+// TODO(windows maintainers)
```
@wrowe for visibility. I think you should probably put a bug # here. No one will know to search for that particular string (whereas sometimes I might grep the codebase for TODO(jmarantz)).

But probably a better pattern is TODO(#xxxx) where xxxx is some issue you add to the issue-tracker for this, and then it can be assigned to whoever.

sriduth replied:
Sorry I Missed that! - Ill create a issue for this rightaway.

sriduth replied:
Created: https://github.com/envoyproxy/envoy/issues/9025

jmarantz commented on source/exe/win32/platform_checks.cc~:
```diff
@@ -0,0 +1,16 @@
+#pragma once
+
+#include "envoy/filesystem/filesystem.h"
```
remove this file from your PR.

sriduth commented:
Ah, I missed a signoff on one commit. Sorry about that! I read somewhere that force pushing makes it harder for review so ill hold off for now.

jmarantz commented on source/exe/posix/platform_checks.cc:
```diff
@@ -0,0 +1,77 @@
+}
+
+// Check system wide configurations such as user level
+// file watches, file limit permissions etc...
```
Move this doc to the header and use doxygen syntax consistent with other Envoy header doc.

jmarantz commented:
If/when you open a new PR to fix DCO I will approve that one & forward to senior maintainers.

jmarantz commented:
/wait

sriduth commented:
Thanks @jmarantz!! I've opened a new PR here: https://github.com/envoyproxy/envoy/pull/9098

sriduth commented:
Closing this PR since a new one is created with DCO fixes.
